src/Security/AuthenticationMiddleware.php

```diff
@@ -60,4 +135,136 @@ public function process(HTTPRequest $request, callable $delegate)
 
         return $delegate($request);
     }
+
+    /**
+     * Check if we have a redirect to a password change form registered
+     * and redirect there if possible.
+     * Otherwise, deauthenticate the user by resetting it for this session,
+     * since we should treat ones with expired passwords as unauthorised.
+     *
+     * @param HTTPRequest $request
+     *
+     * @return HTTPResponse|null
+     */
+    protected function handleExpiredPassword(HTTPRequest $request)
+    {
+        $session = $request->getSession();
+
+        $sessionRedirectUrl = $session->get(static::SESSION_KEY_EXPIRED_PASSWORD_REDIRECT);
+        $defaultRedirectUrl = static::config()->get('expiredPasswordDefaultRedirect');
+
+        if ($sessionRedirectUrl || $defaultRedirectUrl) {
+            $redirectUrl = Director::absoluteURL($sessionRedirectUrl ?? $defaultRedirectUrl);
+        } else {
+            $redirectUrl = null;
+        }
+
+        if ($session && $redirectUrl) {
+            $whitelist = static::config()->get('expiredPasswordWhitelistedUrls');
+            if (!is_array($whitelist)) {
+                $whitelist = [];
+            }
+
+            $sessionWhitelist = $session->get(static::SESSION_KEY_EXPIRED_PASSWORD_WHITELIST);
+            if (is_array($sessionWhitelist)) {
+                $whitelist = array_merge($whitelist, $sessionWhitelist);
+            }
+
+            $currentUrl = Director::absoluteURL($request->getURL(true));
+
+            $allowedUrls = array_merge([
+                $redirectUrl
+            ], array_map(
+                function ($url) {
+                    return Director::absoluteURL($url);
+                },
+                $whitelist
+            ));
+
+            if (!in_array($currentUrl, $allowedUrls, true)) {
+                return $this->redirectOrForbid($request, $redirectUrl);
+            }
+        } else {
+            Security::setCurrentUser(null);
```

Comment:
Is there a reason why you've chosen `Security::setCurrentUser()` (this request) over `IdentityStore::logout` (all requests)?

Reply:
Discussed this one in person. The main reasons are:
   - change password form requires to know what member it changes password for (some form of authentication)
   - there are other possible implementations, but those may require introducing new API for authentication which is different from the standard `Security::getCurrentUser`, which feels to be less ideal.
   - `ChangePasswordHandler` aproach with AutoLoginHash doesn't seem to be applicable here since it merely moves an authentication factor into a GET parameter (token) and then we'd have to manually authenticate user anyway for the change password form to be working

In general:
Logging out here doesn't feel to be giving any benefits, since we'll have to authenticate user later for the change password form to work. Substituting `Security::getCurrentUser` API with something different doesn't feel right either.